### DISCUSSION TITLE
Curvature at LE not monotonous descending

### DISCUSSION THREAD

**Discussion by goney3** [jxjo/Xoptfoil2#7] [Q&A] [answered]
Hello! I have recently been trying to run a NACA 23112 through Xoptfoil2 and I have been able to run the camb-thick and hicks-henne shaping functions, but for some reason when trying Bezier it errors out with _"Top: Curvature at LE not monotonous descending"._ I don't know what this means or how to correct it. I've tried using different bezier counts but no luck. I couldn't find anything in the documentation either. Any advice greatly appreciated, thank you for your time!

![image_2024-06-07_092810090](https://github.com/jxjo/Xoptfoil2/assets/19561317/66a416e4-2667-476f-b4cc-e006f2e57e4b)
Version: 1.0.2 win

**Comment by jxjo** [comment]
Hello Keith, 

... thanks for your feedback!

Mmh - the NACA 23112 is a difficult case ;-) In the lower nose area there is a quite strong bump, which brings to 'Match_Bezier' algorithm to its limits when trying approximate the lower side contour line and at the same time trying to achieve a curvature a leading edge which is equal to the upper side. 
The other 'problem' is, that the curvature on the lower side at leading edge doesn't have its maximum at leading (which is normal) but a little bit behind LE. Zooming in with the AirfoilEditor shows the issue:

![grafik](https://github.com/jxjo/Xoptfoil2/assets/58261241/757997fe-bcfe-4d01-a169-295150605b11)

This is the error message...
But the surrender of the 'Match_Bezier' is the root cause. I will have a look, if I can improve the error handling in this case. 

With the AirfoilEditor I matched a Bezier airfoil with 5 control points on the upper side and 7 on the lower side. Not perfect but I think good enough to start an optimization.

![grafik](https://github.com/jxjo/Xoptfoil2/assets/58261241/c280fb05-79d6-4b23-8300-931c1ea5a76a)

Please find the airfoils in the zip file:

[NACA 23112-bezier.zip](https://github.com/user-attachments/files/15743546/NACA.23112-bezier.zip)

best regards

Jochen

**Reply by goney3** [reply]
Thank you Jochen, though it looks like Xoptfoil2 still doesn't like it as an input, it even "not monotonous descending" even for hicks-henne now, I really do appreciate you looking into this as well as making such incredible programs that hobbyists can use. All the gratitude my friend!

**Reply by jxjo** [reply]
Hello Keith, 
... did you use the file NACA23112-bezier.bez as the seed airfoil? 
(in this case, there will be now 'Match_Bezier' at the beginning of the optimization, as the initial Bezier-Airfoil is already availabale - and the optimization should run) 
regards
Jochen

**Reply by goney3** [reply]
Ah, and that's where I screwed up! Thank you!

**Reply by goney3** [reply]
![image](https://github.com/jxjo/Xoptfoil2/assets/19561317/172de603-392b-4113-9193-da5010fd1f9a)
I was feeling confident that setting the .bez you made as the input would have fixed the issue, but apparently not :'(
I apologize for presenting this headache.

**Reply by jxjo** [reply]
... oh, no! ;-) Strange ...
Could you please send me the input file and the seed airfoil you used?
Thank you for your patience ...

**Reply by goney3** [reply]
I used the JX-GP_bezier input file as my starting point and then tried to squish the Cm as my targets:

!
! Generating NACA 23112 airfoil (similar) - for a better Tandem Wing
!   

&optimization_options
  airfoil_file     = 'NACA23112.bez' 
  shape_functions  = 'bezier'						! hicks-henne, bezier, camb-thick
  cpu_threads      = -1								! take one cpu core less than available 
  show_details     = .true.                         ! show details of operating points and more during optimization
/  

&bezier_options
   ncp_top = 8 
   ncp_bot = 8 
/

&particle_swarm_options
  convergence_profile = 'exhaustive'                   	! 'quick' for fast convergence (results) 
/

&operating_conditions

  dynamic_weighting      = .true.                   ! activate dynamic weighting during optimization
  allow_improved_target  = .true.                   ! results are allowed to be better than target 


  re_default       = 2000000							! use this Reynolds number for operating points

  noppoint = 9										! xfoil will have some work with 11 operating points ...

  op_point(1) = -0.05								! negative cl oppoints for a fine bottom side of airfoil 	
  optimization_type(1) = 'target-drag'
  target_value(1) = 0.0060  			 
  weighting(1) = -0.5  								! not too important - negative value to exclude from dynamic weighting 

  op_point(2) = 0.05   
  optimization_type(2) = 'target-moment'
  target_value(2) = 0.0000  
  weighting(2) = 2.0

  op_point(3) = 0.20
  optimization_type(3) = 'target-moment'
  target_value(3) = 0.0000   
  weighting(3) = 1.0

  op_point(4) = 0.30
  optimization_type(4) = 'target-moment'
  target_value(4) = 0.0000 
  weighting(4) = 1.0

  op_point(5) = 0.50			
  optimization_type(5) = 'target-moment'
  target_value(5) = 0.0000 
  weighting(5) = 1.0

  op_point(6) = 0.65
  optimization_type(6) = 'target-moment'
  target_value(6) = 0.0000  
  weighting(6) = 1.0

  op_point(7) = 0.80
  optimization_type(7) = 'target-moment'
  target_value(7) = 0.0000
  weighting(7) = 1.0

  op_point(8) = 1.00
  optimization_type(8) = 'target-moment'
  target_value(8) = 0.0000   	 
  weighting(8) = 1.0
 
  op_point(9) = 11.9									! ensure good cl-max value
  op_mode(9) = 'spec-al'								! always use 'spec-al' at high cl values!
  optimization_type(9) = 'target-lift'
  target_value(9) = 1.9
  weighting(9) = 0.5
/

&curvature
  auto_curvature = .true.						  ! auto curvature which exceeds or matches
!  max_curv_reverse_top = 1                       ! max no of curvature reversals - top ("reflexed"?)
!  max_curv_reverse_bot = 2                       ! max no of curvature reversals - bot ("rearloading"?)
!  curv_threshold   = 0.1                         ! threshold to detect reversals 
!  max_te_curvature = 5.0                         ! max curvature at trailing edge
!  max_le_curvature_diff = 5.0  
/

&geometry_targets
  ngeo_targets    = 2
  target_type(1)  = 'Camber'
  target_value(1) = 0.0200							! we want an airfoil family with ths camber 
  weighting(1) = 0.5
  target_type(2)  = 'Thickness'
  target_value(2) = 0.18
  weighting(2) = 0.5
/  

&xfoil_run_options
  ncrit = 7												! default value for op points (more turbulentf flow) 
/

**Reply by goney3** [reply]
The hicks-henne created this nifty airfoil though... so even though the bezier wasn't wanting to run, I am happy with this output :)
[NACA23112_Opt5.txt](https://github.com/user-attachments/files/15791522/NACA23112_Opt5.txt)

**Reply by goney3** [reply]
I just tried using an Eppler 334 Airfoil and I get the same error as the NACA 23112
![image](https://github.com/jxjo/Xoptfoil2/assets/19561317/17687e7a-e03e-4737-ae4b-378eec350898)

```
!
! Generating NACA 23112 airfoil (similar) - for a better Tandem Wing
!   
!   	Bezier curves with 5 control points are used as shape functions 
!		The airfoil has rear-loading (one reversal on bottom side) 

&optimization_options
  airfoil_file     = 'E334.bez' 
  shape_functions  = 'bezier'						! hicks-henne, bezier, camb-thick
  cpu_threads      = -1								! take one cpu core less than available 
  show_details     = .false.                         ! show details of operating points and more during optimization
/  

&bezier_options
   ncp_top = 8 
   ncp_bot = 8 
/

&particle_swarm_options
  convergence_profile = 'exhaustive'                   	! 'quick' for fast convergence (results) 
/

&operating_conditions

  dynamic_weighting      = .true.                   ! activate dynamic weighting during optimization
  allow_improved_target  = .true.                   ! results are allowed to be better than target 


  re_default       = 130000							! use this Reynolds number for operating points

  noppoint = 9										! xfoil will have some work with 11 operating points ...

  op_point(1) = -0.05								! negative cl oppoints for a fine bottom side of airfoil 	
  optimization_type(1) = 'target-drag'
  target_value(1) = 0.0060  			 
  weighting(1) = -0.5  								! not too important - negative value to exclude from dynamic weighting 

  op_point(2) = 0.05   
  optimization_type(2) = 'target-moment'
  target_value(2) = 0.0000  
  weighting(2) = 1.0

  op_point(3) = 0.20
  optimization_type(3) = 'target-moment'
  target_value(3) = 0.0000   
  weighting(3) = 1.0

  op_point(4) = 0.30
  optimization_type(4) = 'target-moment'
  target_value(4) = 0.0000 
  weighting(4) = 1.0

  op_point(5) = 0.50			
  optimization_type(5) = 'target-moment'
  target_value(5) = 0.0000 
  weighting(5) = 1.0

  op_point(6) = 0.65
  optimization_type(6) = 'target-moment'
  target_value(6) = 0.0000  
  weighting(6) = 1.0

  op_point(7) = 0.80
  optimization_type(7) = 'target-moment'
  target_value(7) = 0.0000
  weighting(7) = 1.0

  op_point(8) = 1.00
  optimization_type(8) = 'target-moment'
  target_value(8) = 0.0000   	 
  weighting(8) = 1.0
 
  op_point(9) = 11.9									! ensure good cl-max value
  op_mode(9) = 'spec-al'								! always use 'spec-al' at high cl values!
  optimization_type(9) = 'target-lift'
  target_value(9) = 1.85
  weighting(9) = 0.5
/

&curvature
  auto_curvature = .true.						  ! auto curvature which exceeds or matches
  max_curv_reverse_top = 1                       ! max no of curvature reversals - top ("reflexed"?)
  max_curv_reverse_bot = 1                       ! max no of curvature reversals - bot ("rearloading"?)
!  curv_threshold   = 0.1                         ! threshold to detect reversals 
!  max_te_curvature = 5.0                         ! max curvature at trailing edge
!  max_le_curvature_diff = 5.0  
/

&geometry_targets
  ngeo_targets    = 2
  target_type(1)  = 'Camber'
  target_value(1) = 0.0200							! we want an airfoil family with ths camber 
  weighting(1) = 2.0
  target_type(2)  = 'Thickness'
  target_value(2) = 0.8
  weighting(2) = 2.0
/  

&xfoil_run_options
  ncrit = 7												! default value for op points (more turbulentf flow) 
/
```

**Reply by goney3** [reply]
[E334-bez.txt](https://github.com/user-attachments/files/15829474/E334-bez.txt)
Trying hicks-henne didn't like it either :(
It's a little odd that it has a "perfect quality" but still throws the error... is it an issue with reflex airfoils?
![image](https://github.com/jxjo/Xoptfoil2/assets/19561317/c8beb6d8-f45e-4cce-a747-50f9c25bc7a8)

**Comment by jxjo** [comment, marked as answer]
Hello Keith, 

... I had a look at the Eppler airfoil:

![grafik](https://github.com/jxjo/Xoptfoil2/assets/58261241/e3f9846a-8e9c-4b89-92fb-51ee3d6de96d)

Mmh - I decided then to implement something like: 

- if the seed airfoil already is not monotonous at leading edge
- then do not make checks for being monotonous during optimization

Here is a beta-Version for you. Could you test it, if it hels in your use case? 

[Xoptfoil2_1.0.3_beta_win.zip](https://github.com/user-attachments/files/15836100/Xoptfoil2_1.0.3_beta_win.zip)

Thank you!
and regards

**Reply by goney3** [reply]
Well, it got further at least :)
I'll have to play around with this some more now that it can at least clear the first hurdle, thank you for the version to test out!
![image](https://github.com/jxjo/Xoptfoil2/assets/19561317/1528af78-0e69-4507-9deb-17dcee14f775)
This is using the same input file as posted bove for the E334.bez

**Reply by jxjo** [reply]
... oh - no! ;-) 
Could you send me the input file and the seed airfoil?

**Reply by goney3** [reply]
Last night I ran some interpolations since I couldn't get Xoptfoil2 to run the not monotonous descending and came up with this 33% interpolation that seems to do okay at Re 130,000 (in XFLR5 at least).
[33_Interpolation.txt](https://github.com/user-attachments/files/15839803/33_Interpolation.txt)

**Reply by goney3** [reply]
Oh sorry, yeah the E334.bez file and input files are above.... I'll reattach them here though :)
[E334-bez.txt](https://github.com/user-attachments/files/15840219/E334-bez.txt)
```

!
! Generating NACA 23112 airfoil (similar) - for a better Tandem Wing
!   
!   	Bezier curves with 5 control points are used as shape functions 
!		The airfoil has rear-loading (one reversal on bottom side) 

&optimization_options
  airfoil_file     = 'E334.bez' 
  shape_functions  = 'bezier'						! hicks-henne, bezier, camb-thick
  cpu_threads      = -1								! take one cpu core less than available 
  show_details     = .false.                         ! show details of operating points and more during optimization
/  

&bezier_options
   ncp_top = 8 
   ncp_bot = 8 
/

&particle_swarm_options
  convergence_profile = 'exhaustive'                   	! 'quick' for fast convergence (results) 
/

&operating_conditions

  dynamic_weighting      = .true.                   ! activate dynamic weighting during optimization
  allow_improved_target  = .true.                   ! results are allowed to be better than target 


  re_default       = 130000							! use this Reynolds number for operating points

  noppoint = 9										! xfoil will have some work with 11 operating points ...

  op_point(1) = -0.05								! negative cl oppoints for a fine bottom side of airfoil 	
  optimization_type(1) = 'target-drag'
  target_value(1) = 0.0060  			 
  weighting(1) = -0.5  								! not too important - negative value to exclude from dynamic weighting 

  op_point(2) = 0.05   
  optimization_type(2) = 'target-moment'
  target_value(2) = 0.0000  
  weighting(2) = 1.0

  op_point(3) = 0.20
  optimization_type(3) = 'target-moment'
  target_value(3) = 0.0000   
  weighting(3) = 1.0

  op_point(4) = 0.30
  optimization_type(4) = 'target-moment'
  target_value(4) = 0.0000 
  weighting(4) = 1.0

  op_point(5) = 0.50			
  optimization_type(5) = 'target-moment'
  target_value(5) = 0.0000 
  weighting(5) = 1.0

  op_point(6) = 0.65
  optimization_type(6) = 'target-moment'
  target_value(6) = 0.0000  
  weighting(6) = 1.0

  op_point(7) = 0.80
  optimization_type(7) = 'target-moment'
  target_value(7) = 0.0000
  weighting(7) = 1.0

  op_point(8) = 1.00
  optimization_type(8) = 'target-moment'
  target_value(8) = 0.0000   	 
  weighting(8) = 1.0
 
  op_point(9) = 11.9									! ensure good cl-max value
  op_mode(9) = 'spec-al'								! always use 'spec-al' at high cl values!
  optimization_type(9) = 'target-lift'
  target_value(9) = 1.85
  weighting(9) = 0.5
/

&curvature
  auto_curvature = .true.						  ! auto curvature which exceeds or matches
  max_curv_reverse_top = 1                       ! max no of curvature reversals - top ("reflexed"?)
  max_curv_reverse_bot = 1                       ! max no of curvature reversals - bot ("rearloading"?)
!  curv_threshold   = 0.1                         ! threshold to detect reversals 
!  max_te_curvature = 5.0                         ! max curvature at trailing edge
!  max_le_curvature_diff = 5.0  
/

&geometry_targets
  ngeo_targets    = 2
  target_type(1)  = 'Camber'
  target_value(1) = 0.0200							! we want an airfoil family with ths camber 
  weighting(1) = 2.0
  target_type(2)  = 'Thickness'
  target_value(2) = 0.8
  weighting(2) = 2.0
/  

&xfoil_run_options
  ncrit = 7												! default value for op points (more turbulentf flow) 
/
```

**Reply by goney3** [reply]
Hmm... using an Re of 135,000 instead of 130,000 seems to work ... interesting! :)

**Comment by jxjo** [comment]
Hello Keith,
... I had a short look in Xflr5 at the E334.bez. The airfoil shows same strange behaviour with a step in the cl(alpha) polar resulting in a a kink in the cd(alpha) polar at approx alpha = -0.5 

The reason could be, that a bubble at leading edge lower side is formed at alpha < -0.5° :  
![grafik](https://github.com/jxjo/Xoptfoil2/assets/58261241/dc20c067-f47a-4b6d-88f4-33a4acc6b5bd)
 
I assume that xfoil could have repeated difficulties with this at certain op point values...

If xfoil doesn't converge, move the operating point a little up or downwards.
Hope this helps!

regards

Jochen
